src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -0,0 +1,348 @@
+            {
+                serviceType = typeof(NullServiceType);
+            }
+
+            if (!_isImmutable)
+            {
+                return _serviceCollection.Any(sd => sd.ServiceType == serviceType);
+            }
+
+            var service = _serviceProvider.GetService(serviceType);
+            return service != null;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes of the instance.
+        /// </summary>
+        /// <param name="disposing">Whether or not the instance is disposing.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        private static Type GetDictionaryType(Type serviceType) => _dictionaryType.MakeGenericType(serviceType);
+
+        private void RemoveContractService(Type serviceType)
+        {
+            var dicType = GetDictionaryType(serviceType);
+            var sd = _serviceCollection.SingleOrDefault(s => s.ServiceType == serviceType);
+            _serviceCollection.Remove(sd);
+        }
+
+        private ContractDictionary GetContractDictionary(Type serviceType, bool createIfNotExists)
+        {
+            var dicType = GetDictionaryType(serviceType);
+
+            if (_isImmutable)
+            {
+                return (ContractDictionary)ServiceProvider.GetService(dicType);
+            }
+
+            var dic = _serviceCollection
+                .Where(sd => sd.ServiceType == dicType)
+                .Select(sd => sd.ImplementationInstance)
+                .Cast<ContractDictionary>()
+                .SingleOrDefault();
+
+            if (createIfNotExists && dic == null)
+            {
+                dic = (ContractDictionary)Activator.CreateInstance(dicType);
+                _serviceCollection.AddSingleton(dicType, dic);
+            }
+
+            return dic;
+        }
+
+        private class ContractDictionary
+        {
+            private readonly ConcurrentDictionary<string, IList<Func<object>>> _dictionary = new ConcurrentDictionary<string, IList<Func<object>>>();
+
+            public bool IsEmpty => _dictionary.Count == 0;
+
+            public bool TryRemoveContract(string contract) =>
+                _dictionary.TryRemove(contract, out var _);
+
+            public Func<object> GetFactory(string contract) =>
+                GetFactories(contract)
+                    .LastOrDefault();
+
+            public IEnumerable<Func<object>> GetFactories(string contract) =>
+                _dictionary.TryGetValue(contract, out var collection)
+                ? collection
+                : Enumerable.Empty<Func<object>>();
+
+            public void AddFactory(string contract, Func<object> factory) =>
+                _dictionary.AddOrUpdate(contract, _ => new List<Func<object>> { factory }, (_, list) =>
+                {
+                    list.Add(factory);
```

Comment:
this might be an issue if the `RemoveLastFactory` creates the dictionary with `default(IList<Func<object>>)`. Probably want to add a null check and create it.

```
list = list ?? new List<Func<object>>();
```

type thing.

Reply:
Nice catch.
Would it make sense to remove the contract from the dic within the `AddOrUpdate` `updateValueFactory` delegate?

Reply:
```c#
public void RemoveLastFactory(string contract) =>
    _dictionary.AddOrUpdate(contract, default(IList<Func<object>>), (_, list) =>
        {
            var lastIndex = list.Count - 1;
            if (lastIndex > 0)
            {
                list.RemoveAt(lastIndex);
            }

            if (list.Count == 0)
            {
                _dictionary.TryRemove(contract, out var _);
            }

            // TODO if list empty remove contract entirely
            // need to find how to atomically update or remove
            // https://github.com/dotnet/corefx/issues/24246
            return list;
        });
```

Reply:
well the issue is the 2nd parameter, it needs something. 

`UpdateOrRemove` would be nice.

The issue we got is we have to do two operations if we want to do that.

Update and then Remove from Dictionary.

Then it becomes a threading issue and a race condition unless we start using our own `locks`. If we start using our own `locks` we loose the nice per bucket locking mechanism that the `ConcurrentDictionary` offers.

So unfortunately we are just going to have to put up with potentially empty values in our dictionary. If we return null if the count == 0 for the `updateValueFactory` list that's fine.

Just means we have to cater for null list entries inside `AddFactory`

Reply:
For the reference
https://github.com/dotnet/corefx/issues/24246

Reply:
when you do `return list;` in your snippet above it would re-create the entry again, so the `TryRemove` is redundant.

Reply:
oh ok, i thought it's a void list going nowhere

Reply:
> So unfortunately we are just going to have to put up with potentially empty values in our dictionary. If we return null if the count == 0 for the `updateValueFactory` list that's fine.
> 
> Just means we have to cater for null list entries inside `AddFactory`

So what's better empty list or null and null check?

Reply:
probably null and null check I'd imagine, less memory allocations potentially.